### PR TITLE
Update semaphore guide

### PR DESCRIPTION
**Pull request by freddygv**
Addresses: #1940

Following the commands provided by the semaphore guide does not lead to the shown outputs, and key steps to set up a semaphore with the HTTP API were missing.

This PR is an overhaul of the guide so that it gets closer to the logic we have in the CLI.

### CODE REVIEW

File: website/source/docs/guides/semaphore.html.md [hunk 1]
```diff
@@ -65,75 +67,106 @@ This is something like:
...
 
 ```text
 {
-    "Limit": 3,
+    "Limit": 2,
     "Holders": {
-        "4ca8e74b-6350-7587-addf-a18084928f3c": true,
-        "adf4238a-882b-9ddc-4a9d-5b6758e4159e": true
+        "<session>"
     }
```
Comment (banks) [inline website/source/docs/guides/semaphore.html.md]: This is a JSON object but you are using it like an array, should it be a map or array?

File: website/source/docs/guides/semaphore.html.md [hunk 2]
```diff
@@ -65,75 +67,106 @@ This is something like:
...
+    "Value": null,
+    "Session": "<session>",
+    "CreateIndex": 897,
+    "ModifyIndex": 897
+  }
+]
+```
+Note that the `Value` we embedded into `<lock>` is Base64 encoded when returned by the API.
+
+When the `<lock>` is read and its `Value` is decoded, we can verify the remote `Limit` agrees with the `Holders` count.
```
Comment (banks) [inline website/source/docs/guides/semaphore.html.md]: remote? not clear what that is referring to.

File: website/source/docs/guides/semaphore.html.md [hunk 3]
```diff
@@ -65,75 +67,106 @@ This is something like:
...
+If this request succeeds with `true`, the contender now holds a slot in the semaphore. 
+If this fails with `false`, then likely there was a race with another contender to acquire the slot.
 
-Note that the session by default makes use of only the gossip failure detector. That
-is, the session is considered held by a node as long as the default Serf health check
-has not declared the node unhealthy. Additional checks can be specified if desired.
+To re-attempt the acquisition, we watch for changes on `<prefix>`. This is because a slot
+may be released, a node may fail, etc. Slot holders must also watch for changes since the
+slot may be released by an operator or automatically released due to a false positive
+in the failure detector.
```
Comment (banks) [inline website/source/docs/guides/semaphore.html.md]: I'd reword this to point out that a holder MUST watch the lock value constantly (i.e. it must be watching _before_ it starts it's critical work/assumes lock is held). You say that but it's second so I almost skipped over it.

There are several things the watch achieves:

 1. If the operator or something changes the lock state then the holder should consider their lock lost
 2. Similar to above but if the failure detector fails for some fuzzy reason (packet loss, CPU overload etc) and the Consul server considers the session closed even though the process is still running, the process needs to know about it.

So the watch is important to guarantee that the process still holds the lock - it should consider it's lock revoked if either the watch fails to connect for a period (while it is retried, lock is lost) or if it gets an update that doesn't have the node in the holders list any more.

File: website/source/docs/guides/semaphore.html.md (hunk 1 above)
Comment (freddygv) [reply website/source/docs/guides/semaphore.html.md]: So in our CLI we use a map as a set [here](https://github.com/hashicorp/consul/blob/7ed3b23333b4ab30aa2774e650bc17c12244cf2d/api/semaphore.go#L87)

I'll change it to an array so users can turn it into a set however they like

File: website/source/docs/guides/semaphore.html.md (hunk 2 above)
Comment (freddygv) [reply website/source/docs/guides/semaphore.html.md]: true, going to remove that word

**Comment by freddygv** [comment]
@topfunky have you had a chance to take a look at this?

**Comment by pearkes** [comment]
@topfunky happy to merge this unless you want 👀on it?
